tls/datadog_checks/tls/tls.py

```diff
@@ -0,0 +1,353 @@
+        else:
+            self.check = self.check_remote
+            self._tags.append('server_hostname:{}'.format(self._server_hostname))
+            self._tags.append('server:{}'.format(self._server))
+            self._tags.append('port:{}'.format(self._port))
+
+        # Assign lazily since these aren't used by both collection methods
+        self._validation_data = None
+        self._local_cert_loader = None
+        self._tls_context = None
+
+    def check_remote(self, instance):
+        if not self._server:
+            raise ConfigurationError('You must specify `server` in your configuration file.')
+
+        try:
+            sock = self.create_connection()
+        except Exception as e:
+            self.service_check(self.SERVICE_CHECK_CAN_CONNECT, self.CRITICAL, tags=self._tags, message=str(e))
+            return
+        else:
+            self.service_check(self.SERVICE_CHECK_CAN_CONNECT, self.OK, tags=self._tags)
+
+        # Get the cert & TLS version from the connection
+        with closing(sock):
+            try:
+                with closing(self.tls_context.wrap_socket(sock, server_hostname=self._server_hostname)) as secure_sock:
+                    der_cert = secure_sock.getpeercert(binary_form=True)
+                    protocol_version = secure_sock.version()
+            except Exception as e:
+                # https://docs.python.org/3/library/ssl.html#ssl.SSLCertVerificationError
+                err_code = getattr(e, 'verify_code', None)
+                message = getattr(e, 'verify_message', str(e))
+                self.service_check(self.SERVICE_CHECK_VALIDATION, self.CRITICAL, tags=self._tags, message=message)
+
+                # There's no sane way to tell it to not validate just the expiration
+                # This only works on Python 3.7+, see: https://bugs.python.org/issue28182
+                # https://github.com/openssl/openssl/blob/0b45d8eec051fd9816b6bf46a975fa461ffc983d/include/openssl/x509_vfy.h#L109
+                if err_code == 10:
+                    self.service_check(
+                        self.SERVICE_CHECK_EXPIRATION, self.CRITICAL, tags=self._tags, message='Certificate has expired'
+                    )
+
+                return
+
+        # Load https://cryptography.io/en/latest/x509/reference/#cryptography.x509.Certificate
+        try:
+            cert = load_der_x509_certificate(der_cert, default_backend())
+        except Exception as e:
+            self.service_check(
+                self.SERVICE_CHECK_VALIDATION,
+                self.CRITICAL,
+                tags=self._tags,
+                message='Unable to parse the certificate: {}'.format(e),
+            )
+            return
+
+        self.check_protocol_version(protocol_version)
+        self.validate_certificate(cert)
+        self.check_age(cert)
+
+    def check_local(self, instance):
+        if not self._local_cert_path:
```

Comment:
Do we need this check? Looks like the only way we get here is if this file specifically exists?